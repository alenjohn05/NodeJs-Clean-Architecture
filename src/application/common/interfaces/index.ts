export * from './ILogger';
export * from './IPostsRepository';
