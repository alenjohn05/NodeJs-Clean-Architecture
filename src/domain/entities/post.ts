export class Post {
  public id?: number;
  public createdAt: Date;
  public published: boolean;
  public title: string;

  constructor(post: { id?: number; createdAt: Date; published: boolean; title: string }) {
    this.id = post.id;
    this.createdAt = post.createdAt;
    this.published = post.published;
    this.title = post.title;
  }
}
