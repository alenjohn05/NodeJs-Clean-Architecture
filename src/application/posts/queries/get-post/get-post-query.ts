import { Dependencies } from '@infrastructure/di';
import { NotFoundException } from '@application/common/exceptions';
import { toDto } from './get-post-query-mapper';

export type GetPostQuery = Readonly<{
  id: number;
}>;

export function makeGetPostQuery({ postsRepository }: Pick<Dependencies, 'postsRepository'>) {
  return async function getPostQuery({ id }: GetPostQuery) {
    const post = await postsRepository.getById({ id });

    if (!post) {
      throw new NotFoundException(`Post ${id} does does not exist`);
    }

    return toDto(post);
  };
}
