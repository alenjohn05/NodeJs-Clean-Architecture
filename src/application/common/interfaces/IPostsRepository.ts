import { Post } from '@domain/entities';

export interface IPostsRepository {
  create(parameters: { post: Post }): Promise<{ id: number }>;
  delete(parameters: { id: number }): Promise<void>;
  getById(parameters: { id: number }): Promise<Post | null>;
  list(parameters: { pageNumber: number; pageSize: number }): Promise<{ count: number; posts: Array<Post> }>;
  update(parameters: { post: Post }): Promise<void>;
}
