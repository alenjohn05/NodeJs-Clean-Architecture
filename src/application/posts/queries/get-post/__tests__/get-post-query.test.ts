import { mockDeep } from 'jest-mock-extended';
import { IPostsRepository } from '@application/common/interfaces';
import { NotFoundException } from '@application/common/exceptions';
import { Post } from '@domain/entities';
import { makeGetPostQuery } from '../get-post-query';

describe('getPostQuery', () => {
  function setup() {
    const postsRepository = mockDeep<IPostsRepository>({
      getById: jest.fn().mockResolvedValue(
        new Post({
          id: 1,
          createdAt: new Date(2022, 1, 1),
          published: false,
          title: 'Mock post',
        }),
      ),
    });

    const getPostQuery = makeGetPostQuery({ postsRepository });

    return {
      getPostQuery,
      postsRepository,
    };
  }

  test('gets post and returns domain transfer object', async () => {
    const { getPostQuery } = setup();

    const result = await getPostQuery({ id: 1 });

    expect(result).toMatchSnapshot();
  });

  describe('given the post does not exist', () => {
    test('throws not found exception', async () => {
      const { getPostQuery, postsRepository } = setup();
      postsRepository.getById.mockResolvedValue(null);

      const result = getPostQuery({ id: 1 });

      await expect(result).rejects.toThrow(NotFoundException);
    });
  });
});
