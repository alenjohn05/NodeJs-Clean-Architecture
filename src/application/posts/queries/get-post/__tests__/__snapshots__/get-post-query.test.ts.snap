// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPostQuery gets post and returns domain transfer object 1`] = `
{
  "id": 1,
  "published": false,
  "title": "Mock post",
}
`;
